src/swarm/neo/util/Batch.d

```diff
@@ -0,0 +1,236 @@
+// TODO
+
+/+
+
+* It should support records with fields of any type. (The old one just supports
+  records of `<char[]>` and `<char[], char[]>`.)
+* It should provide a method to get the batched size of a record. (The old one
+  has this function internally, but it's private.)
+* Compression should be optional.
+* The maximum batch size (== the record size limit) should be soft and
+  configurable at runtime, not a compile-time constant. This way, the size limit
+  could be set in the node's config file, which could then be re-parsed if the
+  limit needed to be changed. Future batches would then use the new size limit.
+* Instead of checking that each added record fits in the remaining space in the
+  batch, check that the size of each added record is <= the maximum batch size.
+  This will lead to flexible sized batches of at most 2x the maximum size. (For
+  the advantage of this approach, see
+  https://github.com/sociomantic-tsunami/swarm/issues/158#issuecomment-339671247.)
+
++/
+
+module swarm.neo.util.Batch;
+
+import ocean.transition;
+import ocean.io.compress.Lzo;
+
+struct BatchWriter
+{
+    import ocean.core.Traits;
+
+    private void[]* buffer;
+
+    // Has two functions:
+    //   1. Size limit per tuple added.
+    //   2. Once a batch exceeds this size, it is considered full.
+    private size_t max_size;
+
+    public void initialise ( void[]* buffer, size_t max_size )
+    {
+        this.buffer = buffer;
+        this.max_size = max_size;
+    }
+
+    public size_t sizeOf ( T ... ) ( T items )
+    {
+        size_t sum;
+        foreach ( item; items )
+            sum += this.sizeOfItem(item);
+
+        return sum;
+    }
+
+    public bool fits ( T ... ) ( T items )
+    {
+        return this.sizeOf(items) <= this.max_size;
+    }
+
+    // Returns: true if added, false if too big
+    public bool add ( T ... ) ( T items, void delegate ( ) send_batch )
+    {
+        if ( !this.fits(items) )
+            return false;
+
+        foreach ( item; items )
+            this.addItem(item);
+
+        if ( this.buffer.length >= this.max_size )
+            send_batch();
```

Comment:
This looks wrong: `send_batch` is called only if `items` precisely fit in the buffer, otherwise one of the `if` conditions yields `false`.

Reply:
Mm, well it's intended to only call `send_batch` if the buffer is full. Seems correct to me...?

Reply:
Suppose `this.max_size == 60`, and you call `add` repeatedly, each time with one item of size 40. On the first call `this.buffer.length >= this.max_size` yields `false` so `send_batch` is not called. On subsequent calls `this.fits(items)` returns `false` so `send_batch` is not called either. So `send_batch` is never called.

Reply:
You're misreading `this.fits`.

Reply:
Oh.

Reply:
LGTM.

Reply:
Suggest raising an error (`assert/verify`) if `this.buffer.length >= this.max_size` and this method is called, or calling `clear` after `send_batch`.

Reply:
Yep, I added an extra commit that clears after sending :+1: